explorer/interpreter/type_checker.cpp

```diff
@@ -4556,6 +4567,22 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
     class_vtable[fun->name()] = {fun, class_level};
   }
 
+  if (class_decl->extensibility() == ClassExtensibility::None) {
+    for (const auto& vt : class_vtable) {
+      const auto* const fun = vt.getValue().first;
+      if (!fun->is_method()) {
+        continue;
+      }
+
+      if (fun->virt_override() == VirtualOverride::Abstract) {
+        return ProgramError(class_decl->source_loc())
+               << "Error declaring `" << class_decl->name() << "`"
+               << ": final class should implement abstract method `"
+               << fun->name() << "`.";
+      }
+    }
+  }
+
```

Comment:
```suggestion
  if (class_decl->extensibility() == ClassExtensibility::None && VTableHasAbstractMethods(class_vtable)) {
    return ProgramError(class_decl->source_loc())
                 << "Error declaring `" << class_decl->name() << "`"
                 << ": final class should implement abstract method `"
                 << fun->name() << "`.";
  }
```

Reply:
Would be reusable here: https://github.com/carbon-language/carbon-lang/pull/2528/files#diff-b7a0337bf9cf26348bc16c40cd497a179d6f3d19862eb689f6238fa288ca347dR1862-R1880